i18n: Add charset related settings

doronbehar opened the pull request:
## Description of changes


## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [x] (Module updates) Added a release notes entry if the change is significant
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

doronbehar committed:
i18n: Add charset related settings

Fixes #404758

FliegendeWurst reviewed (commented):
What happens if I set this?

```
i18n.extraLocaleSettings.LC_TIME = "de_DE.UTF-8";
i18n.localeCharsets.LC_TIME = "ISO-8859-15";
```

As far as I can tell it won't fail eval, but it will probably fail to apply properly at runtime.

doronbehar committed:
nixosTests.i18n: init

doronbehar commented:
> What happens if I set this?
> 
> ```
> i18n.extraLocaleSettings.LC_TIME = "de_DE.UTF-8";
> i18n.localeCharsets.LC_TIME = "ISO-8859-15";
> ```
> 
> As far as I can tell it won't fail eval, but it will probably fail to apply properly at runtime.

Nice example. It correctly fails (build, not eval) with:

```
       > Error: unsupported locales detected:
       > de_DE.UTF-8/ISO-8859-15 \
```

Whereas this doesn't fail:

```nix
        i18n = {
          defaultLocale = "en_US.UTF-8";
          extraLocaleSettings = {
            LC_TIME = "de_DE";
          };
          localeCharsets = {
            LC_TIME = "ISO-8859-1";
          };
        };
```

NOTE also that I had to switch to `8859-1` != `8859-15`. The above is now part of `nixosTests.i18n`.

doronbehar committed:
nixos/i18n: Remove special handling of LANGUAGE

Partially revert #179194 which first filtered
`extraLocaleSettings.LANGUAGE`. Indeed this environment variable is
given precedence according to:

https://www.gnu.org/software/gettext/manual/html_node/Locale-Environment-Variables.html

So using it will simply make all other `extraLocaleSettings` be ignored
in practice, but the OS shouldn't ignore it when aggregating the locales
required in general, as this setup should still be legitimate, and it
may even be useful if you wish to set it by default and use `env -u
LANGUAGE` for some programs.

FliegendeWurst commented on nixos/modules/config/i18n.nix:
```diff
@@ -5,15 +5,24 @@
+    ]
+    ++ lib.pipe config.i18n.extraLocaleSettings [
+      # TODO: Explain why is this filter added here...
+      (lib.filterAttrs (n: v: n != "LANGUAGE"))
```
$LANGUAGE specifies the preferred languages by the user, e.g. `de:en_US`

In my opinion it would be more robust to fail evaluation if `i18n.extraLocaleSettings.X` is set, and X doesn't start with `LC_`.

doronbehar replied:
> $LANGUAGE specifies the preferred languages by the user, e.g. `de:en_US`
> 
> In my opinion it would be more robust to fail evaluation if `i18n.extraLocaleSettings.X` is set, and X doesn't start with `LC_`.

I investigated this a bit, and added a 3rd commit addressing and explaining my opinion.

FliegendeWurst reviewed (commented):
Change looks okay. Needs a changelog entry (i18n.extraLocales usage changed).

doronbehar commented:
> Needs a changelog entry (i18n.extraLocales usage changed).

I agree! I will do it in a few hours, hopefully after @musjj , @jansol and/or @eclairevoyant will raise their voice too.

jansol commented:
The documentation definitely an improvement, but I'm not convinced of this approach.

To clarify, my goal is to have multiple locales generated, even though I don't intend to use them by default. I mostly want to have them around for a few cases of legacy software that does manual string operations with certain assumptions about locale and charset. (i.e. I would be running individual scripts/binaries with e.g. `LANG=de_DE.ISO-8859-15 legacy-app` while keeping the default as UTF-8)

doronbehar commented:
> Needs a changelog entry (i18n.extraLocales usage changed).

Done!

doronbehar commented:
> To clarify, my goal is to have multiple locales generated, even though I don't intend to use them by default. I mostly want to have them around for a few cases of legacy software that does manual string operations with certain assumptions about locale and charset. (i.e. I would be running individual scripts/binaries with e.g. `LANG=de_DE.ISO-8859-15 legacy-app` while keeping the default as UTF-8)

@jansol please share with us the full details of the locale settings you wish to set. I believe your usecase should be covered with the changes here - the `extraLocales` option should now allow you to install `de_DE/ISO-8859-1`, without putting it in `extraLocaleSettings`.

doronbehar commented:
Thanks for the review @FliegendeWurst . Both comments were addressed. I think this should be ready now.

jansol commented:
> @jansol please share with us the full details of the locale settings you wish to set. I believe your usecase should be covered with the changes here

Right, I misread the diff & got sidetracked by the split extraLocaleSettings/localeCharsets tests.

What I want to do is basically
```nix
i18n = {
  defaultLocale = "en_US.UTF-8";
  extraLocales = [
    "en_US.UTF-8/UTF-8"
    "de_DE.UTF-8/UTF-8"
    "de_DE/ISO-8859-1"
  ];
};
```

Which does indeed seem to be possible now.

SuperSandro2000 commented:
> Nice example. It correctly fails (build, not eval) with:

Can we easily let that fail with an assertion?

> Needs a changelog entry (i18n.extraLocales usage changed).

I think that was missed. I needed to dig a little to know that I needed to remove .UTF-8 from my extraLocaleSettings or add a /UTF-8

doronbehar commented:
> > Needs a changelog entry (i18n.extraLocales usage changed).

The `i18n.extraLocales` setting is a completely new setting for this NixOS release, so the fact the details of it's usage were changed after a period of <month it was available in `nixos-unstable` doesn't justify editing that changelog line.
